### DISCUSSION TITLE
deal with non-prismatic beam in xc

### DISCUSSION THREAD

**Discussion by ebrahimraeyat** [xcfem/xc#72] [Q&A]
Hi All. I wanted to work with xc and for starting i want to  model an industrial structure with XC, my question:

can I model non-prismatic beam in xc. I saw this discution that  seems it is not directly possible in opensees:

[How_to_deal_with_non-prismatic_beam_in_Opensees](https://www.researchgate.net/post/How_to_deal_with_non-prismatic_beam_in_Opensees)

I have been modeled it in [ETABS v19](https://www.csiamerica.com/products/etabs) and wanted to model it with xc and compare both of them:

![image](https://user-images.githubusercontent.com/8196112/103029500-2d61c280-456f-11eb-8873-2b7cbd5f7897.png)

**Comment by lcpt** [comment]
Hi @ebrahimraeyat 

You can model the tapered beams discretizing its length (that's certainly the solution that ETABS uses). All the solutions proposed in your ResearchGate link seem adequate. On the other hand, if you need a very precise solution (for example in the context of a PhD dissertation, it's better to use shell or brick elements).

**Comment by ebrahimraeyat** [comment]
Thanks, you mean we don't have an element for non-prismatic at all (in theoretical and thus in none of FEM software). Thanks.
I want it for my engineering purpose.

**Comment by lcpt** [comment]
Well, as Mr. Balduzzi says, there are more accurate elements like the one in [this article](https://www.researchgate.net/publication/295396523_Non-prismatic_Beams_a_Simple_and_Effective_Timoshenko-like_Model). I you want to implement those elements in XC (or OpenSees of both), I'll be happy to help.

On the other hand, for practical purposes, I would use the discretization mentioned before for the model of the whole structure, and make a more precise analysis when designing the connections by means of a CBFEM model.

**Reply by ebrahimraeyat** [reply]
> Well, as Mr. Balduzzi says, there are more accurate elements like the one in [this article](https://www.researchgate.net/publication/295396523_Non-prismatic_Beams_a_Simple_and_Effective_Timoshenko-like_Model). I you want to implement those elements in XC (or OpenSees of both), I'll be happy to help.

Unfortunately I don't have work  experience with c++, but if I can develop it with python I will be happy to apply it in XC

**Comment by ebrahimraeyat** [comment]
Is it possible to share some of these models?

**Reply by lcpt** [reply]
The code that generates those models is not ready to be published yet. If you want to analyze how it works I can share the code privately. Ask me using l.pereztato@ciccp.es.

**Comment by ebrahimraeyat** [comment]
Hi @lcpt, How can I model general Steel I/Wide Flange sections, I mean without manually calculate the properties of sections like A,  Iz or Ix?
thanks

**Comment by lcpt** [comment]
Hi @ebrahimraeyat 

You don't need to manually calculate the section properties you can use the steel shape database (at present AISC and Arcelor). You can also compute automatically the mass properties of any polygon using the geom module. Furthermore, you can also [model the steel shape using fiber sections]( https://portwooddigital.com/2020/05/10/meshing-for-column-loads) if you want to reproduce the nonlinear behavior of the material.

**Comment by ebrahimraeyat** [comment]
Thanks @lcpt. I don't want to model standard I/Wide Flange. I wanted to model general, I mean whatever BF, TF, H and TW. is it possible?

**Reply by lcpt** [reply]
Why it will not be possible? The solutions I mentioned earlier are completely general.

**Comment by ebrahimraeyat** [comment]
Thanks, @lcpt. I don't want to perform a nonlinear analyzing, can you write it on one line? For example this is in examples:

`shape= ASTM_materials.WShape(steel,'W18X50')`

but I wanted to set (BF, TF, D, TW), not general polygon:

`shape= ASTM_materials.WShape(steel,(20, 12, 30, 8))  # (BF, TF, D, TW)`

**Reply by ebrahimraeyat** [reply]
Excuse me, I found it on materials/sections/section_properties.py. thanks.

```
from materials.sections import section_properties as sp
wdTopFlange = wdBotFlange = 20
thTopFlange = thBotFlange = 1.2
thWeb = .8
hgWeb = 15

isection = sp.ISection("isection",wdTopFlange,thTopFlange,thWeb,hgWeb,wdBotFlange,thBotFlange)
```

**Comment by ebrahimraeyat** [comment]
I created two functions for creating lines and section for nonprismatic section:


```python
import xc_base
import geom
import xc

from materials.sections import section_properties
from materials import typical_materials as tm
```

creating lines giving 2 points

```python
def create_lines(
    p1,
    p2,
    n: int,
    points,
    lines,
    point_id,
    )-> list:
    x1, y1, z1 = p1.getPos.x, p1.getPos.y, p1.getPos.z
    x2, y2, z2 = p2.getPos.x, p2.getPos.y, p2.getPos.z
    dx = (x2 - x1) / n
    dy = (y2 - y1) / n
    dz = (z2 - z1) / n
    
    first_point = p1
    new_lines = []
    for i in range(n):
        point_id += 1
        x = first_point.getPos.x + dx
        y = first_point.getPos.y + dy
        z = first_point.getPos.z + dz
        if all((x == x2, y == y2, z == z2)):
            last_point = p2
        else:
            last_point = points.newPntIDPos3d(point_id, geom.Pos3d(x, y, z))
        l = lines.newLine(first_point.tag, last_point.tag)
        new_lines.append(l)
        first_point = last_point
    return new_lines, point_id
```
creating sections, but I don't know how can I apply section on one line, not on Set:

```python
def apply_nonPrismatic_section(sets, bf, tf, tw, hw_min, hw_max, \
                               coordTransf, modelSpace, material, prep):
    tag = 'isection'
    isection = section_properties.ISection(tag,bf,tf,tw,hw_min,bf,tf)
    linel_mat= tm.BeamMaterialData(name=tag, section=isection, material=material)
    linel_mat.setupElasticShear3DSection(preprocessor=prep)
    for Set in sets:
        tag = Set.getLines[0].tag
        # n = Set.lines.size
        # dhw = hw_max - hw_min
        # for line in Set.lines:
        #     z = line.getCentroid().z
        #     average_hw = hw_min + z / h_col * dhw
        #     isection = section_properties.ISection(f"isection{line.tag}",bf,tf,tw,average_hw,bf,tf)
        #     p1 = line.getP1().tag
        #     p2 = line.getP2().tag
        #     seedElemHandler.defaultMaterial = isection.sectionName
        #     seedElemHandler.newElement('elastic_beam3d', xc.ID([p1, p2]))
        modelSpace.createElasticBeams(Set, linel_mat, coordTransf, xc.Vector([0.0,1.0,0.0]), nDiv= 6)
```
creaitng lines:

```python
col_frame_a, n = create_lines(pt0, pt3, 10, points,lines, n)
col_frame_c, n = create_lines(pt1, pt7, 10, points,lines, n)
col_frame_e, n = create_lines(pt2, pt11, 10, points,lines, n)
```

assign sections:
```python
apply_nonPrismatic_section([col_frame_a_set, col_frame_c_set], bf_col_ae, \
                           tf_col_ae, tw_col_ae, hw_col_ae_bot, hw_col_ae_top, coordTransf, \
                           modelSpace, steel_W, preprocessor)
```
Thanks.

complete code:

[sole_zeinali.zip](https://github.com/xcfem/xc/files/5988090/sole_zeinali.zip)

**Reply by lcpt** [reply]
Hi.

Materials (a section is a material) are assigned to elements when the element is created. You specify the material template to use when the `genMesh()` method is called using the `defaultMaterial`property of the seed element handler (see, for example, `test_uls_checking_aisc_01.py`). You can store the material name you'll like to use in the lines or anywhere else (you can use the `setProp()` and `getProp()` methods to store things in the C++ object that lies inside the Python wrapper provided by the boost.python interface).

The element material can also be modified after creation, but I don't use this very often.

The methods of the `ModelSpace` class are mostly convenience functions to make easier simple tasks. You can take a look inside to see how the task is performed when you need to do things that are not so simple.

**Comment by ebrahimraeyat** [comment]
Thanks, @lcpt. I wrote the function (assume developing line in z direction, sufficient for my case now!):

```python

def apply_nonPrismatic_section(sets, bf, tf, tw, hw_min, hw_max, \
                               material, prep, element_handler):
    # dia_bounds = {
    #     'x': ['dia_bound.x', 'l_center.x', 'bnd.getZMin'],
    #     'y': ['dia_bound.y', 'l_center.y', 'bnd.getZMin'],
    #     'z': ['dia_bound.z', 'l_center.z', 'bnd.getZMin'],
    # }
    dhw = hw_max - hw_min
    for Set in sets:
        bnd = Set.lines.getBnd()
        dz = bnd.diagonal.z
        zmin = bnd.getZMin
        for line in Set.getLines:
            z_line = line.getCentroid().z
            percent = abs(z_line - zmin) / dz

            average_hw = hw_min + percent * dhw
            assert average_hw > 0
            # lines_len += l
            isection = section_properties.ISection(f"isection{line.tag}",bf,tf,tw,average_hw,bf,tf)
            xc_material= isection.defElasticShearSection3d(prep, material)
            element_handler.defaultMaterial = xc_material.name
            element_handler.newElement('ElasticBeam3d', xc.ID([0,0]))
            line.genMesh(xc.meshDir.I)
            for e in line.getElements:
                e.setProp('crossSection',isection)
        Set.fillDownwards()
```

But when  I want to create a report, It gives me this error for elements that using ISection shape class:

`DL
Load combination: DL already exists, redefined.
XC::ElasticBeam3d::getPyProp; Warning, property: 'ULS_normalStressesResistanceSect1' not found. Returning None.
Traceback (most recent call last):
  File "/home/ebi/xc/my_models/sole_zeinali/xc_write_reports.py", line 66, in <module>
    report.checksReport(limitStateLabel,[],[],setsBmEl,argsBmEl)
  File "/usr/local/lib/python3.8/dist-packages/postprocess/reports/report_generator.py", line 69, in checksReport
    self.displayBeamResult(attributeName=limitStateLabel,itemToDisp=argS,beamSetDispRes=stV,setToDisplay=stV,caption=capt,fileName=fullgrFileNm+'.jpg')
  File "/usr/local/lib/python3.8/dist-packages/postprocess/output_handler.py", line 779, in displayBeamResult
    lstArgVal=[e.getProp(attributeName+'Sect1')(itemToDisp) for e in beamSetDispRes.elements]
  File "/usr/local/lib/python3.8/dist-packages/postprocess/output_handler.py", line 779, in <listcomp>
    lstArgVal=[e.getProp(attributeName+'Sect1')(itemToDisp) for e in beamSetDispRes.elements]
TypeError: 'NoneType' object is not callable`

But for other elements works fine. 
```python
allMemberSet = modelSpace.setSum('allMemberSet', [
                                #  beam1Set,
                                 beam2Set,
                                #  col_frame_a_set,
                                #  col_frame_e_set,
                                 col_frame_c_set,
                                 ])
```
The commented sets are sets that created with ISection shape.

**Reply by lcpt** [reply]
Hi @ebrahimraeyat 

It seems that the problem is due to the fact, that the `ISection` class has not its counterparts in ASTM_materials.py and `aisc_metric_shapes.py` (see `WShape`, `CShape`, etc. in both files). Those classes implement AISC code checking for the standard shapes. To make this work with ISection we need to implement the corresponding AISC checking for «generic» I-Shaped sections.

**Comment by ebrahimraeyat** [comment]
Thanks @lcpt. Maybe inheritance and then write special method for built-up I-shaped sections? Thanks.

Is it better to open an issue?

**Comment by ebrahimraeyat** [comment]
Implemented as per commit #88 and here is an example:

https://github.com/ebrahimraeyat/xc_soleh
